astropy/cosmology/parameter.py

```diff
@@ -0,0 +1,265 @@
+            The output of calling ``fvalidate(cosmology, self, value)``
+            (yes, that parameter order).
+        """
+        return self.fvalidate(cosmology, self, value)
+
+    @classmethod
+    def register_validator(cls, key, fvalidate=None):
+        """Decorator to register a new kind of validator function.
+
+        Parameters
+        ----------
+        key : str
+        fvalidate : callable[[object, object, Any], Any] or None, optional
+            Value validation function.
+
+        Returns
+        -------
+        ``validator`` or callable[``validator``]
+            if validator is None returns a function that takes and registers a
+            validator. This allows ``register_validator`` to be used as a
+            decorator.
+        """
+        if key in cls._registry_validators:
+            raise KeyError(f"validator {key!r} already registered with Parameter.")
+
+        # fvalidate directly passed
+        if fvalidate is not None:
+            cls._registry_validators[key] = fvalidate
+            return fvalidate
+
+        # for use as a decorator
+        def register(fvalidate):
+            """Register validator function.
+
+            Parameters
+            ----------
+            fvalidate : callable[[object, object, Any], Any]
+                Validation function.
+
+            Returns
+            -------
+            ``validator``
+            """
+            cls._registry_validators[key] = fvalidate
+            return fvalidate
+
+        return register
+
+    # -------------------------------------------
+
+    def __repr__(self):
+        return f"<Parameter {self._attr_name!r} at {hex(id(self))}>"
+
+
+# ===================================================================
+# Built-in validators
+
+
+@Parameter.register_validator("default")
+def _validate_with_unit(cosmology, param, value):
+    """
+    Default Parameter value validator.
+    Adds/converts units if Parameter has a unit.
+    """
+    if param.unit is not None:
+        with u.add_enabled_equivalencies(param.equivalencies):
+            value = u.Quantity(value, param.unit)
+    return value
+
+
+@Parameter.register_validator("float")
+def _validate_to_float(cosmology, param, value):
+    """Parameter value validator with units, and converted to float."""
+    value = _validate_with_unit(cosmology, param, value)
+    return float(value)
+
+
+@Parameter.register_validator("scalar")
+def _validate_to_scalar(cosmology, param, value):
+    """"""
+    value = _validate_with_unit(cosmology, param, value)
+    if not value.isscalar:
+        raise ValueError(f"{param.name} is a non-scalar quantity")
+    return value
+
+
+@Parameter.register_validator("non-negative")
+def _validate_non_negative(cosmology, param, value):
+    """Parameter value validator where value is a positive float."""
+    value = _validate_to_float(cosmology, param, value)
+    if value < 0.0:
+        raise ValueError(f"{param.name} can not be negative.")
```

Comment:
Total nitpick: "can not" -> "cannot"

Reply:
SGTM. I think that's legacy, but I'm changing the intro to ``{param.name}``, so it's fine to change other things.

Reply:
Done!